bundles/org.eclipse.equinox.p2.artifact.repository/src/org/eclipse/equinox/internal/p2/artifact/repository/simple/DownloadJob.java

```diff
@@ -56,8 +56,7 @@ protected IStatus run(IProgressMonitor jobMonitor) {
 			if (masterMonitor.isCanceled())
 				return Status.CANCEL_STATUS;
 			// process the actual request
-			SubMonitor subMonitor = SubMonitor.convert(masterMonitor.slice(1), 1);
-			IStatus status = repository.getArtifact(request, subMonitor);
+			IStatus status = repository.getArtifact(request, masterMonitor.newChild(1));
```

Comment:
Please note that submonitor is not threadsafe! Also now the job monitor do not recive any updates.

It might therefore be better to pass in an own implementation that:

1. syncronize on the master before update it
2. also send all events to the job monitor

Reply:
I'm trying to understand the nature of the comment.  Even before the change that introduced problem (call to slice) it was a submonitor already then too:

![image](https://github.com/user-attachments/assets/f1f26d8f-b1a8-4817-94d6-d0bab21c5567)

Was your comment intended for the change in the other file?

https://github.com/eclipse-equinox/p2/pull/586/files#diff-9a34bd9b50e863d2faa5988b40185db6a3a7e144ef93175808e9e44c19394840

Note that I did test that I see logged information in the installer with these changes, so I'm not sure why the job monitor is not receiving updates...

Maybe we shouldn't be changing this at all?   It's very surprising to me that every change has a different gotcha.  It's still not clear if anything was actually improved...

Reply:
Note that the code I changed was a botched migration to the SubMonitor, which started this whole ordeal. Originally, the now deprecated SubProgressMonitor was used.

![image](https://github.com/user-attachments/assets/1c517c37-91a7-4c48-a729-eacd527505f7)

https://github.com/eclipse-equinox/p2/commit/9009085b72e9507fdc4b7d0adb87883adefc5637#diff-98743f36a52a7f919e0b5cb2ce651e1a89f1f3a7abf1c343e58a29ee9340e2b5

> syncronize on the master before update it

Without synchronization, some ticks might indeed get lost due to parallelism. But I wonder if this is really worth the additional effort... `done()` is explicitly called on the SubMonitor, so any remaining ticks will be consumed once the download is done, so it's at least not a critical problem.

> also send all events to the job monitor

Wouldn't this effectively mean both jobs show the same task?

Reply:
If the assertion "The subMonitor created in the run() method is only used on this job's thread, i.e., on the thread that is calling run()" a true assertion?  I thought was was the case, but I've been known to be wrong.

I would have though only the thread safety of the masterMonitor is a potential concern...

Reply:
@ptziegler has correctly explained it sorry for being late.

One should also keep in mind that newChild/split clear the work acquired by a previous call, so if you do:

```
IProgressMonitor m1 = monitor.split(10);
IProgressMonitor m2 = monitor.split(10);
```

then you have one Monitor (`m1`) that is already done, so only the last job started could report actual progress!

Reporting good and consistent progress is a very hard topic indeed and Eclipse makes its even harder some times due to historical reasons.

The "slice" was introduced by me a while ago specifically for the case where one do work in parallel Threads (usually streams or jobs) that still perform a common task but want to report overall progress. Adapted to this situation here for example one would have one (of unknown origin) Monitor and set its task to "Downloading files", and the amount of work to the number of files. Now I start a thread for each file and pass it a slice of one and this thread can now use safely the monitor (but needs to call done possibly) to report its overall download progress while everything is aggregated into the original one, that might observe the number of running Threads and update the message (e.g. with "Downloading files 5 / 10").

Another reason for slice is to notify the main monitor about that progress was distributed (what is impossible with SubMonitor at the moment) so an UI can show something like this:

![grafik](https://github.com/user-attachments/assets/a7fad8dd-7258-40d6-b878-2e007a02d733)
 _Image from [here](https://stackoverflow.com/questions/2270863/displaying-progressbar-for-file-upload/2311153)_